### PR TITLE
Add critical kinds to crd store

### PR DESCRIPTION
**Pull request by bianpengyuan**
Check that all critical kinds are ready when initializing crd store. Currently critical kinds include rule, template, instance, handler, adapter and attributesmanifest. Address #8839

### CODE REVIEW

File: mixer/pkg/config/crd/store.go [hunk 1]
```diff
@@ -146,12 +150,26 @@ func (s *Store) Init(kinds []string) error {
 	s.informers = make(map[string]cache.SharedInformer, len(kinds))
 	remaining := s.checkAndCreateCaches(d, lwBuilder, kinds)
 	if len(remaining) > 0 {
+		for _, k := range remaining {
+			if s.isCriticalKind(k) {
```
Comment (mandarjog) [inline mixer/pkg/config/crd/store.go]: Please add a configurable timeout and an active poll for missing critical kinds, giving up so early is actually expensive. Default waiting for upto 10s is reasonable.

File: mixer/pkg/config/crd/store.go [hunk 2]
```diff
@@ -146,12 +150,26 @@ func (s *Store) Init(kinds []string) error {
...
+				panic(fmt.Errorf("critical kind %s is not ready", k))
+			}
+		}
 		log.Warnf("Failed to discover kinds: %v", remaining)
 	}
```
Comment (mandarjog) [inline mixer/pkg/config/crd/store.go]: If we fail to discover these `kinds` then we should do
`go func(remaining int) {
   for remaining > 0 { 
           sleep(sleeptime)
           remaining = s.ckeckAndCreateCaches()
   }
}(remaining)

File: mixer/pkg/config/crd/store.go [hunk 3]
```diff
@@ -146,12 +150,26 @@ func (s *Store) Init(kinds []string) error {
 	s.informers = make(map[string]cache.SharedInformer, len(kinds))
 	remaining := s.checkAndCreateCaches(d, lwBuilder, kinds)
 	if len(remaining) > 0 {
+		for _, k := range remaining {
+			if s.isCriticalKind(k) {
+				panic(fmt.Errorf("critical kind %s is not ready", k))
```
[the remark this answers is not in the document]
Comment (bianpengyuan) [reply mixer/pkg/config/crd/store.go]: Done. Changed it to return error, which essentially will make mixer server fail to start up.

File: mixer/pkg/config/crd/store.go (hunk 1 above)
Comment (bianpengyuan) [reply mixer/pkg/config/crd/store.go]: Done. Use the retryTimeout and retryInterval that are already existed. tetryTimeout is configurable through URL query param.

File: mixer/pkg/config/crd/store.go (hunk 2 above)
Comment (bianpengyuan) [reply mixer/pkg/config/crd/store.go]: Added retry. Done.

File: mixer/pkg/config/crd/store.go [hunk 4]
```diff
@@ -145,13 +153,41 @@ func (s *Store) Init(kinds []string) error {
 	s.caches = make(map[string]cache.Store, len(kinds))
 	s.informers = make(map[string]cache.SharedInformer, len(kinds))
 	remaining := s.checkAndCreateCaches(d, lwBuilder, kinds)
+	timeout := time.After(s.retryTimeout)
+	tick := time.Tick(s.retryInterval)
+	stopRetry := false
+	if tick != nil {
```
Comment (mandarjog) [inline mixer/pkg/config/crd/store.go]: when will tick be `nil` it is initialized 2 lines above that.

File: mixer/pkg/config/crd/init.go [hunk 5]
```diff
@@ -62,7 +62,7 @@ func (b *dynamicListerWatcherBuilder) build(res metav1.APIResource) cache.Lister
 }
 
 // NewStore creates a new Store instance.
-func NewStore(u *url.URL, gv *schema.GroupVersion) (store.Backend, error) {
+func NewStore(u *url.URL, gv *schema.GroupVersion, ck []string) (store.Backend, error) {
 	kubeconfig := u.Path
 	namespaces := u.Query().Get("ns")
 	retryTimeout := crdRetryTimeout
```
Comment (mandarjog) [inline mixer/pkg/config/crd/init.go]: // k8s://?retry-timeout=1m

this should be `u.Query().get("retry-timeout")` right?

File: mixer/pkg/config/crd/store.go (hunk 4 above)
Comment (bianpengyuan) [reply mixer/pkg/config/crd/store.go]: Right, it is not necessary. Initially I thought retryInterval should also be configurable and tick will be nil if interval is configured as 0. But now it is not configurable so it should be safe to assume tick cannot be nil.

File: mixer/pkg/config/crd/init.go (hunk 5 above)
Comment (bianpengyuan) [reply mixer/pkg/config/crd/init.go]: Yes, it is extracted in the lines below.

File: mixer/pkg/config/crd/store.go [hunk 6]
```diff
@@ -145,13 +153,39 @@ func (s *Store) Init(kinds []string) error {
 	s.caches = make(map[string]cache.Store, len(kinds))
 	s.informers = make(map[string]cache.SharedInformer, len(kinds))
 	remaining := s.checkAndCreateCaches(d, lwBuilder, kinds)
+	timeout := time.After(s.retryTimeout)
+	tick := time.Tick(s.retryInterval)
```
Comment (mandarjog) [inline mixer/pkg/config/crd/store.go]:
```go
defer tick.Stop()
```

Comment (mandarjog) [reply mixer/pkg/config/crd/store.go]: nevermind, it is a channel, not a Ticker.

**Comment by hklai** [comment]
@bianpengyuan Thanks. Please cherry-pick it back to master if it is needed for 1.1.

**Pull request merged**